### PR TITLE
AN-6745: Migration between app versions freezes the app - Part 2.

### PR DESCRIPTION
**Pull request by BradleyWire**
## What's new in this PR?

Part 2 of https://github.com/wireapp/wire-android/pull/2679. 

Added Test cases for: 

- Cache entry migration from version 24 to 25
- ActiveAccountsDao
- CacheEntryDao
- TeamsDao
#### APK
[Download build #1529](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1529/artifact/build/artifact/wire-dev-PR2685-1529.apk)
[Download build #1530](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1530/artifact/build/artifact/wire-dev-PR2685-1530.apk)
[Download build #1531](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1531/artifact/build/artifact/wire-dev-PR2685-1531.apk)
[Download build #1533](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1533/artifact/build/artifact/wire-dev-PR2685-1533.apk)
[Download build #1534](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1534/artifact/build/artifact/wire-dev-PR2685-1534.apk)
[Download build #1540](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1540/artifact/build/artifact/wire-dev-PR2685-1540.apk)
[Download build #1541](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1541/artifact/build/artifact/wire-dev-PR2685-1541.apk)

### CODE REVIEW

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    private fun createActiveAccountsList() =
+        listOf(
+            createActiveAccount(TEST_ACTIVE_ACCOUNT_ID_ACTIVE),
+            createActiveAccount(TEST_ACTIVE_ACCOUNT_ID_INACTIVE)
+        )
+
+    @After
+    fun tearDown() {
+        globalDatabase.close()
+    }
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt]: suggestion: I like to keep it just below @Before method

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/CacheEntryDaoTest.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        val cachedEntries = getListOfCacheEntries()
+        cachedEntries.map {
+            cacheEntryDao.insertCacheEntry(it)
+        }
+
+        val roomActiveAccounts = cacheEntryDao.cacheEntries()
+        assert(roomActiveAccounts[0].key == TEST_CACHE_ENTRY_FIRST_ID)
+        assert(roomActiveAccounts[1].key == TEST_CACHE_ENTRY_SECOND_ID)
+        roomActiveAccounts.map {
+            assert(roomActiveAccounts.size == 2)
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/CacheEntryDaoTest.kt]: can be moved out of "map"

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/TeamsDaoTest.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        val teams = getListOfTeams()
+        teams.map {
+            teamsDao.insertTeam(it)
+        }
+
+        val roomActiveAccounts = teamsDao.allTeams()
+        assert(roomActiveAccounts[0].teamId == TEST_TEAM_FIRST_ID)
+        assert(roomActiveAccounts[1].teamId == TEST_TEAM_SECOND_ID)
+        roomActiveAccounts.map {
+            assert(roomActiveAccounts.size == 2)
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/TeamsDaoTest.kt]: same as above

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    fun setup() {
+        globalDatabase = Room.inMemoryDatabaseBuilder(
+            getApplicationContext(),
+            GlobalDatabase::class.java
+        ).build()
+        activeAccountsDao = globalDatabase.activeAccountsDao()
+    }
+
+    @Test
+    fun givenAccessTokenIsCalled_whenUserIdIsActiveUser_thenDataShouldBeTheSame() = runBlocking {
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt]: about naming, I see given/when/then case as, "given preconditions, when an action is made, then we see the outcome". so, imo, the method names in tests should be in the form of: `givenAnActiveAccount_whenAccessTokenIsCalledWithItsId_thenDataShouldBeSame`
(wording can be changed)

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+import com.waz.zclient.storage.db.accountdata.ActiveAccountsEntity
+import com.waz.zclient.storage.db.accountdata.SsoIdEntity
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+
+@ExperimentalCoroutinesApi
+class ActiveAccountsDaoTest : IntegrationTest() {
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt]: These tests seem like general purpose dao tests, rather than migration related tests. No need to separate them now but I feel like they could be in a separate pr.

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/globaldatabase/GlobalDatabase24to25MigrationTest.kt [hunk 6]
```diff
@@ -79,6 +80,40 @@ class GlobalDatabase24to25MigrationTest : IntegrationTest() {
...
+            val cachedEntry = cachedEntries[0]
+            assert(cachedEntry.key == TEST_CACHE_ENTRY_ID)
+            assert(cachedEntry.fileId == TEST_CACHE_ENTRY_FILE_ID)
+            assertNull(cachedEntry.data)
+            assert(cachedEntry.lastUsed == TEST_CACHE_ENTRY_LAST_USED)
+            assert(cachedEntry.timeout == TEST_CACHE_ENTRY_TIME_OUT)
+            assert(cachedEntry.filePath == TEST_CACHE_ENTRY_FILE_PATH)
+            assert(cachedEntry.mime == TEST_CACHE_ENTRY_MIME)
+            assert(cachedEntry.encKey == TEST_CACHE_ENTRY_ENC_KEY)
+            assert(cachedEntry.length == TEST_CACHE_ENTRY_LENGTH)
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/globaldatabase/GlobalDatabase24to25MigrationTest.kt]: could be 
```
with(getCachedEntries[0]) {
    assert(key == TEST_CACHE_ENTRY_ID),
    assert(fileId == TEST_CACHE_ENTRY_FILE_ID),
   ...
}
```
for readability.

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/globaldatabase/GlobalDatabase24to25MigrationTest.kt [hunk 7]
```diff
@@ -110,13 +148,23 @@ class GlobalDatabase24to25MigrationTest : IntegrationTest() {
...
+        //Cache Entry
+        private const val TEST_CACHE_ENTRY_ID = "1"
+        private const val TEST_CACHE_ENTRY_FILE_ID = "fileId"
+        private const val TEST_CACHE_ENTRY_LAST_USED = 38847746L
+        private const val TEST_CACHE_ENTRY_TIME_OUT = 1582896705028L
+        private const val TEST_CACHE_ENTRY_FILE_PATH = "/data/downloads/"
+        private const val TEST_CACHE_ENTRY_FILE_NAME = "cachentry2"
+        private const val TEST_CACHE_ENTRY_MIME = ".txt"
+        private const val TEST_CACHE_ENTRY_ENC_KEY = "AES256"
+        private const val TEST_CACHE_ENTRY_LENGTH = 200L
```
Comment (gizemb) [inline storage/src/androidTest/kotlin/com/waz/zclient/storage/globaldatabase/GlobalDatabase24to25MigrationTest.kt]: can they be scoped to calling method rather than being class constants here?

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt (hunk 1 above)
Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt]: Seem relevant to the idea that we should group the functions by similarity moreover logically in Kotlin. Old Java habits.

File: storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt (hunk 5 above)
Comment (BradleyWire) [reply storage/src/androidTest/kotlin/com/waz/zclient/storage/accountdata/ActiveAccountsDaoTest.kt]: I'll bear that in mind for future PRs. But, seemed reasonable to get them done whilst we were heavily testing the migrations
